Add TextStripMarkerDecoder to support <nowiki></nowiki>, refs #232

mwjames opened the pull request:
This PR is made in reference to: #232

This PR addresses or contains:

- Adds `$smwgDecodeTextAnnotationWithStripMarker` setting (disabled by default) to support annotations like:
  - `[[Has description::<pre>sample pre text</pre>]]`
  - `[[Has description::<nowiki>{{#ask: HasStripMarkers }}</nowiki>]]`  
  - `[[Has description::<code><nowiki>{{#ask: HasStripMarkers }}</nowiki></code>]]`
  - Decoding of `<ref>...</rev>` is not supported
- Only supported for text type values

This PR includes:
- [x] Tests (unit/integration)
- [x] CI build passed

Fixes #232

mwjames commented:
Need more tests (especially integration tests)

WolfgangFahl commented:
@mwjames this would be just great if it worked. I tried it out but it doesn't have to seeme the effect I expected. I used:
```{{#set:testCode=<pre>Some new Text</pre>}}```
as a test with a Property testCode of type Text. 
I have set:
```
$GLOBALS['smwgEnabledAuxillaryTextDecoder'] = true;
```
after manually applying this patch on a test wiki.
How could I debug this?

mwjames commented:
> Need more tests (especially integration tests)

I just added some integration tests [0] that should cover most common use cases but `TextStripMarkerDecoderTest` needs extension.

> expected/need. I used:
{{#set:testCode=<pre>Some new Text</pre>}}
as a test with a Property testCode of type Text. 

This will not work with `#set` because the change is primarily focused on the `InTextAnnotationParser` (== `[[ ... :: ...]]`). The processing of `[[ ... :: ...]]` and `#set` are separate and distinct from each other in terms of when and how the parser is invoked.

I just rebased this PR to make it pass the current master with some additional tests that I had written in April but never felt too strongly in pushing it again.

> How could I debug this?

Best the use the fresh rebase with the cited [0] examples.

If you want to make it work with `#set` [1] you should inject `TextStripMarkerDecoder` and do a similar manipulation as added by this PR to the `InTextAnnotationParser`.

[0] tests/phpunit/Integration/ByJsonScript/Fixtures/p-0409 intext nowiki unstrip.json
[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/includes/parserhooks/SetParserFunction.php

WolfgangFahl commented:
I got the pull request and the tests working in a docker environment as well as natively on MacOSX / macports. See https://www.semantic-mediawiki.org/wiki/User:WolfgangFahl/Workdocumentation_2015-12-25_WF

the next step is to add more tests and then implement the missing functionality.

WolfgangFahl commented:
It took a few days to get my enviroment up and running see #1333.
The Test I am trying to implement in SetParserFunctionTest.php is:
        /**
	 * test that <tag></tag> content can be parsed e.g. <pre>Some text</pre>
	 * @since 2.4
	 * @author Wolfgang Fahl
	 */
	public function testTagParsing() {
		$property = new DIProperty( 'BarText' );
		$property->setPropertyTypeId( '_txt' );
		$params = array( 'Foo=sometext','BarText=<pre>Some text</pre>');
		$expected = array(
			'errors' => 0,
			'propertyCount'  => 2,
			'propertyLabels' => array( 'Foo','BarText' ),
			'propertyValues' => array( 'Sometext','<pre>Some text</pre>' )
		);
		$this->testInstantiatedPropertyValues($params,$expected);		
	}

I assume I need to register the property so that BarText will be a "known" property of Type text. How is this done? With the current code above the functon parserUserValue in SMW_DV_WikiPage is called and that's not the one that needs to be fixed.

mwjames commented:
@s7eph4n This should support `<pre>` as part of a text annotation (incl. `#set`, `#subobject`).

WolfgangFahl commented:
Thanx for working on this. Since my comments are gone and the issue is still open I am interested to learn more about the status of things. Will this be part of the next release?

mwjames commented:
@kghbln FYI

> Will this be part of the next release?

Yes.

kghbln commented:
[Documented on wiki](https://www.semantic-mediawiki.org/w/index.php?title=Help%3A%24smwgDecodeTextAnnotationWithStripMarker&type=revision&diff=57757&oldid=56803)

mwjames commented:
> Documented on wiki

`smwgDecodeTextAnnotationWithStripMarker` was merged into `smwgParserFeatures` (#2802).

kghbln commented:
Thanks for the note. I have not worked my way through. Probably later this evening ...
